### ISSUE TITLE
./bootstrap.sh in cabal/cabal-install relies on nonportable features.

### ISSUE THREAD

**Issue by R030t1** [haskell/cabal#4640]
[This line](https://github.com/haskell/cabal/blob/master/cabal-install/bootstrap.sh#L1) will prevent Bash or other extended features from being used on some systems regardless of whether the script ends up being run with Bash or not.

Indeed, if I do not replace `/bin/sh` with `/bin/bash` on Kubuntu 17.07 I receive:

```
./bootstrap.sh: 27: ./bootstrap.sh: ghc: not found
```

Likely because line 27 makes use of a nonportable feature.

**Comment by 23Skidoo** [comment]
The intention is to have it work with any POSIX shell. We should make the Travis check more strict.

**Comment by ezyang** [comment]
@R030t1 The link you posted is missing a line number. Also, line 27 seems to run correctly on posh for me. What is non-compliant?

**Comment by 23Skidoo** [comment]
What I find strange here is that `/bin/sh` links to `dash` on Ubuntu, and `bootstrap.sh` works correctly on Travis.

Code in question also works for me locally in `/bin/sh` on an Ubuntu 16.04 machine:

```
$ /bin/sh
$ GHC="${GHC:-ghc}"
$ echo $GHC
ghc
$ GHC_VER="$(${GHC} --numeric-version)"
$ echo $GHC_VER
8.2.1
```

So it's most likely a misconfiguration issue, I think that @R030t1 has added `ghc` to `PATH` in `.bashrc` instead of `.profile`. Closing as invalid, but feel free to reopen if you disagree.

**Comment by R030t1** [comment]
@ezyang The line number given is 1 which I meant to reference, my apologies if I also should have linked line 27.

@23Skidoo I set my `PATH` in `.bashrc` and have since added it to `.profile`. `bootstrap.sh` should inherit a `PATH` with `ghc` in it regardless of the interpreter used (as far as I am aware; environment variables are part of the POSIX call to launch a process and every program inherits an environment). If I do not change the interpreter to `bash` I receive the error given when running the file as it is cloned from the repository.

Maybe the title is wrong, but I still think there is a bug. I am using a fresh Kubuntu 17.04 installation running in a VM. I untarred the GHC 8.2.1 release to `~/.local` and added that directory to my path, now we are here.

I apologize as this may be related to the other bug I reported - if I freshly clone the repository, changing the interpreter of `bootstrap.sh` will compile some packages successfully, otherwise a lot of packages fail saying that Setup.hs can't find ghc. I am unsure of what information to provide, but I will respond to the other report tomorrow.
